auth 'goToEdit' missing

mantis-toboggan-md opened the pull request:
- #2273 I think this is related to AllowedPrincipalIds being undefined; that shouldn't happen if the auth config flow completes correctly (looking at you saml), but I added a catch for the possibility anyway.
- Also, some small changes to ldap inputs.
- Somehow my goToEdit button made it into master but not the goToEdit function. So here is that function!

mantis-toboggan-md commented:
Per some feedback from @richard-cox, I've updated this PR to correct behavior with the cancel button:
- if configuring/enabling a new provider, cancel returns to the provider-selection screen
- if editing  (aka re-configuring and enabling) a currently-enabled provider, cancel returns to the add users/groups view for that provider
- no cancel button for the add users/groups view

mantis-toboggan-md commented:
Updated to reset any changes made to the locally-stored model object when 'cancel' is used on the edit config page.

richard-cox commented on mixins/auth-config.js:
```diff
@@ -130,10 +141,6 @@ export default {
         await this.reloadModel();
         this.isEnabling = false;
         btnCb(true);
-        if ( wasEnabled ) {
```
On /c/local/auth/config/keycloak?mode=edit` the user isn't taken back to the original view. I think adding `this.editConfig = false;` will fix this`

mantis-toboggan-md replied:
You're completely right; added

richard-cox reviewed (changes requested):
There's one inline comment, apart from that there's some issues when viewing the auth providers via the cluster explorer (`c/local/explorer/management.cattle.io.authconfig`). Think all of these could be `mode`/`as` related
- This is where the misalignment of the principal 'remove` buttons are
- Clicking on `Edit Config` button in the banner results in the page showing inputs that are all disabled
- Clicking on three dot button --> `Edit Config` results in normal edit page showing .. just without top right config/yaml toggle

mantis-toboggan-md commented:
@richard-cox  you're right about the weirdness of `c/local/explorer/management.cattle.io.authconfig/<id>` being related to the `mode` query param. When navigating to `/auth/config/<id>` the mode is always set to 'edit' but that isn't true for the auto-generated list view from the explorer. Since none of the components using the `auth-config` mixin are intended for any mode other than 'edit' I've added a little check in that mixin to apply that query param if not present. I think that's probably the most robust way of avoiding this issue. I considered removing management.cattle.io.authconfig from the explorer nav altogether but I think the list view there has utility if a user has more than one provider enabled.

On a similar note, I updated the cancel button logic to return to the management.cattle.io.authconfig list if that's where the user navigated from, rather than sending them to the provider-picker view in `/auth/config`. 

I'm adding @codyrancher as a reviewer now as I've touched some of the azuread code, most particularly the `test` vuex action used by googleoauth, azuread, github, & saml

richard-cox commented on mixins/auth-config.js:
```diff
@@ -115,10 +137,7 @@ export default {
-              code:   obj.code,
-              config: obj
-            }, { redirectUnauthorized: false });
+            await this.model.doAction('testAndApply', obj, { redirectUnauthorized: false });
```
Setting up github initially works fine, however when editing the provider with the same details it errors with `server error while authenticating: Received Error from github bad_verification_code, description from github The code passed is incorrect or expired.` Might be caused here by the `code` being dropped.

mantis-toboggan-md replied:
I don't remember my reasoning for adding `&& !wasEnabled` [here](https://github.com/rancher/dashboard/pull/2309/files/18ed3c9b87856ccc9d62a3f35e6d1d381181799d#diff-b3838ef0862d5e9f887dcd18548f5233e0453b47455b8963fde00b4b5289169dR122) but it was bad. Removing this check allows the oauth config flow to re-run on edit/save.

codyrancher replied:
This change here is what was causing the `server error while authenticating: parameter 'activeDirectoryEndpoint' cannot be empty` error.

The backend seems to expect the config to be nested under config and not flat at least for azuread.

mantis-toboggan-md replied:
With my correction to line 121 this morning from `if (configType === 'oauth' && !wasEnabled)` -> `if (configType === 'oauth')` this line shouldn't ever be hit by enabling/editing azureAD, as it needs to follow the oauth route (`auth/test` vuex action).
I'm a little surprised `testAndApply` is an available action for this provider actually, since (assuming I read your code correctly yesterday) it's not needed to enable auth.

codyrancher replied:
testAndApply is what actually enables auth for azureAD. I tested with the code you updated this morning and modified this line to fix it.

mantis-toboggan-md replied:
Hmm I misread my own code how supremely embarrassing. Well, modifying that call to testAndApply would interfere github, google, and ldap so I altered azuredad.vue `toSave` to nest the model under a  `config` key (github does the same though the key is `githubConfig`)

richard-cox replied:
Tested editing both github and saml, both successfully saved

richard-cox reviewed (changes requested):
I've tested with Github and SAML and the previous issues are resolved. I did spot a couple of new things though, one inline to do with editing Github and the one below. 

- The 'Loading' indicator appears and is not remove when clicking on `Auth Provider` menu item in side nav whilst already on the auth provider page

mantis-toboggan-md commented:
> * The 'Loading' indicator appears and is not remove when clicking on `Auth Provider` menu item in side nav whilst already on the auth provider page

This is kinda interesting, it only happens when a provider is already enabled so I have to think it's related to [this redirect](https://github.com/rancher/dashboard/blob/master/pages/c/_cluster/auth/config/index.vue#L16); however, I wasn't able to isolate the problem beyond that just poking around this morning. While I'm perfectly willing to investigate this bug next, I think it's outside the scope of this PR (the bug is present on master), which is already substantial; saml and ldap providers at least are broken in master right now and QA is blocked. I would rather not continue to hold them back over this.

codyrancher commented:
With these changes I'm getting an error with the AuzreAD auth. I'm taking a look and trying to figure out what it is.

![image](https://user-images.githubusercontent.com/55104481/107800989-be0d8100-6d1c-11eb-8b26-95011c7856b1.png)

codyrancher reviewed (approved):
It works for AzureAD now. I verified that I can create, edit and disable. Once Richard approves I think it's safe to merge.

richard-cox reviewed (commented):
I've given everything another test and stumbled on something I may have missed before. Editing an enabled github provider incorrectly shows the `GitHub is currently disabled.` banner. 
As the banner's section now covers cases other than `model.enabled` I think the banner just needs a `v-if="!model.enabled". I've tested that and it works fine, applying the same to other provider components should work too? There's some where this isn't an issue (there's no disabled banner at all).

mantis-toboggan-md commented:
>  I think the banner just needs a `v-if="!model.enabled". I've tested that and it works fine, applying the same to other provider components should work too?

You're right; I added this check so the banner only shows if the provider is disabled, and added the banner to some of the edit pages which were missing it previously

richard-cox reviewed (approved):
LGTM
